webservice/webservice.py

```diff
@@ -322,7 +326,64 @@ def get_manifest():
     # Get the response back
     logger.info("Creating the API response")
     response = es_td.transform_manifest(filters=filters)
-    # Return the excel file
+    if output_formats == 'tsv':  # returns metadata as TSV file
+        return response
+    elif output_formats == 'bdbag':  # returns presigned URL to a BDBag in S3
+        response_bag = bdbag_response(response)
+        return jsonify(response_bag)
+
+
+def bdbag_response(response_obj):
+    logger = logging.getLogger('dashboardService.webservice.bdbag_response')
+    # Create and return the BDbag folder.
+    bag_name = 'manifest'
+    bag_info = {'organization': 'UCSC Genomics Institute',
+                'data_type': 'TOPMed',
+                'date_created': datetime.datetime.now().isoformat()}
+
+    # Instantiate bag object.
+    bag = BagHandler(data=response_obj.get_data(),
+                     bag_info=bag_info,
+                     bag_name=bag_name)
+    zipped_bag = bag.create_bag()  # path to compressed bag
+    logger.info("Creating a compressed BDbag containing manifest.")
+
+    # Import bucket environment variable, launch instance of S3-file handler,
+    # and upload the BDBag file to S3.
+
+    # Transfer parameters.
+    aws_region = 'us-west-2'
+    bucket_key = str(uuid.uuid4())
+    azul_s3_bucket = os.getenv("AZUL_S3_BUCKET")
+    access_key_id = os.getenv("AWS_ACCESS_KEY_ID")
+    secret_key = os.getenv("AWS_SECRET_ACCESS_KEY")
+
+    s3 = S3FileHandler(aws_region, access_key_id, secret_key)
+    r = s3.upload_object_to_bucket(azul_s3_bucket,
+                                   zipped_bag,
+                                   bucket_key)
+    os.remove(zipped_bag)
+
+    if r['status_code'] == 200:
+        logger.info("Uploaded BDbag {} to S3 bucket {}.".format(bucket_key,
+                                                                azul_s3_bucket))
+        # Generate presigned URL to access that s3 location.
+        response = s3.create_presigned_url(azul_s3_bucket, bucket_key)
+        response['s3_upload_status_code'] = r['status_code']
+        if response['status_code'] == 200:
+            logger.info("Successfully created presigned URL for bucket {}.".
+                        format(''.join([azul_s3_bucket, '/', bucket_key])))
+        else:
+            logger.debug("Failed to create presigned URL for bucket {}.".
+                        format(''.join([azul_s3_bucket, '/', bucket_key])))
+    else:
+        logger.error("Upload to S3 bucket {} failed: {}".format(
+            azul_s3_bucket, r['status_code']))
+        response = {
+            'presigned_url': None,
```

Comment:
If there is an error, your HTTP response status code needs to indicate the error. If I'm reading this correctly, you will return a 200 status code with the error status code in the body. If that is the case, please return a 4xx or 5xx as appropriate.

I don't think you should have presigned_url in the response body. The response of an error does not have to match the response body of success.

See these links for some ideas: 

https://medium.com/@suhas_chatekar/return-well-formed-error-responses-from-your-rest-apis-956b5275948

https://apigee.com/about/blog/technology/restful-api-design-what-about-errors (although don't do it the Facebook way from this link).

Reply:
Pl. see my response to the above comment.

Reply:
Also, some unit testing with mocking for the webservice would really help...